File: Hayaa.Seed/Hayaa.ProgrameSeed/Component/ProgramDistributedConfig.cs
using Hayaa.ProgrameSeed.Config;
using Hayaa.ProgrameSeed.Model.Config;
using Hayaa.ProgrameSeed.Util;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Web;

namespace Hayaa.ProgrameSeed
{
    class ProgramDistributedConfig
    {
        private AppConfig _appConfig;
        private SeedConfig _seedConfig;
        private static ProgramDistributedConfig _instance = new ProgramDistributedConfig();

        internal static ProgramDistributedConfig Instance
        {
            get
            {
                return _instance;
            }


        }

        private ProgramDistributedConfig()
        {
            _appConfig = new AppConfig()
            {
                Components = new List<ComponentConfig>(),
                ConfigContent = "",
                SolutionID = Guid.Empty,               
                SolutionName = "本地配置",
                Version = 0,
            };//构造函数默认对象内属性数值，默认为本地模式参数
            try
            {
                _seedConfig = ReadSeedConfig();//防止基础配置不存在或者错误导致程序无法启动
            }
            catch (Exception ex)
            {
                _seedConfig = new SeedConfig() { Version = 0, AppConfigID = Guid.Empty, IsRemote = false };//错误配置下给予最小化配置               
            }
        }

        /// <summary>
        /// 本地配置模式下只有一个方案序列化文件
        /// </summary>
        /// <param name="seedConfig"></param>
        private void ReadLocal(SeedConfig seedConfig, InitResult result)
        {
            if (string.IsNullOrEmpty(seedConfig.LocalConfigDirectoryPath))//如果没有默认路径不读取
            {
                result.Result = false;
                result.Message = "本地配置文件路径为空";
                return;//构造函数里默认数值
            }
            try
            {
                var temp = Newtonsoft.Json.JsonConvert.DeserializeObject<AppConfig>(File.ReadAllText(seedConfig.LocalConfigDirectoryPath + "/" + seedConfig.AppConfigFileName));
                if (temp != null)//使用构造函数里的数值，避免多位置同效代码赋值
                {
                    _appConfig = temp;
                }
            }
            catch (Exception ex)//预期异常：格式错误，错误内容
            {
                result.Result = false;
                result.Message = ex.Message;
                return;
            }

        }
        /// <summary>
        /// 读取程序的配置
        /// </summary>
        /// <param name="seedConfig"></param>
        /// <returns></returns>
        private AppConfig ReadLocal(SeedConfig seedConfig)
        {
            if (string.IsNullOrEmpty(seedConfig.LocalConfigDirectoryPath))//如果没有默认路径和默认文件配置不读取
            {
                return null;//构造函数里默认数值
            }
            try
            {
                var temp = Newtonsoft.Json.JsonConvert.DeserializeObject<AppConfig>(File.ReadAllText(seedConfig.LocalConfigDirectoryPath + "/" + seedConfig.AppConfigFileName));
                return temp;
            }
            catch (Exception ex)//预期异常：格式错误，错误内容
            {

            }
            return null;
        }
        /// <summary>
        /// 远程获取程序配置
        /// </summary>
        /// <param name="seedConfig"></param>
        private void ReadRemote(SeedConfig seedConfig)
        {

           
                AppConfig localconfig = null;
            //判断配置文件是否已经存在
            if (File.Exists(seedConfig.LocalConfigDirectoryPath + "/app.Config"))
            {
                localconfig = ReadLocal(seedConfig);
            }
            
                //远程拉取配置文件
                var remoteConfig = GetRemote(seedConfig.SeedServerUrl, seedConfig.AppConfigID);
                //判断配置文件的新鲜程度
                if (remoteConfig != null)//无法获取远程配置时不更新本地
                {
                    if (seedConfig.Version == 0)//永远最新
                    {
                        File.Delete(seedConfig.LocalConfigDirectoryPath + "/" + seedConfig.AppConfigFileName);
                        //固化指定目录下制定的文件
                        File.AppendAllText(seedConfig.LocalConfigDirectoryPath + "/" + seedConfig.AppConfigFileName, Newtonsoft.Json.JsonConvert.SerializeObject(remoteConfig));
                     }
                    if ((seedConfig.Version > 0) && (localconfig == null))//本地没有配置文件并且不是永远更新
                    {
                        File.AppendAllText(seedConfig.LocalConfigDirectoryPath + "/" + seedConfig.AppConfigFileName, Newtonsoft.Json.JsonConvert.SerializeObject(remoteConfig));
                    }
            }
            
        }
        private AppConfig GetRemote(string url, Guid solutionID)
        {
            var dic = new Dictionary<string, string>();
            dic.Add(SeedDefineTable.UrlAction, SeedDefineTable.GetRmoteConfigAction);
            dic.Add(SeedDefineTable.SolutionIDParam, solutionID.ToString());
            string str = "";
            AppConfig result = null;
            try
            {
                /// var apiStoreUser = SecurityProvider.GetApiStoreUser();
                str = HttpRequestHelper.Instance.GetNormalRequestResult(url, dic);
                str = HttpUtility.UrlDecode(str);
                result = Newtonsoft.Json.JsonConvert.DeserializeObject<AppConfig>(str);
            }
            catch (Exception ex)
            {
                result = null;
                return result;
            }
            return result;
        }
        /// <summary>
        /// 
        /// </summary>
        /// <returns></returns>
        private SeedConfig ReadSeedConfig()
        {
            SeedConfig r = new SeedConfig()
            {
                 IsRemote = true,
                IsVirtualPath = true,
                LocalConfigDirectoryPath = "~/Config",
                Version = 0
            };
            try
            {
                string baseDirectory = AppDomain.CurrentDomain.BaseDirectory;//获取根目录路径
                r=Newtonsoft.Json.JsonConvert.DeserializeObject<SeedConfig>(File.ReadAllText(baseDirectory + "/seed.config"));//读取根目录下的配置文件
            }
            catch (Exception ex)
            {
                r = new SeedConfig() { IsVirtualPath = false };
            }
            if (r.IsVirtualPath)//web系统相对路径则获取绝对路径
            {
                r.LocalConfigDirectoryPath= System.Web.HttpContext.Current.Server.MapPath(r.LocalConfigDirectoryPath);
            }
            return r;
        }
        public AppConfig GetLocalConfig()
        {
            return ReadLocal(_seedConfig);
        }
        public bool IsEmpty()
        {
            if (_appConfig.SolutionID.Equals(Guid.Empty))
            {
                return true;
            }
            return false;
        }
        public bool IsFactory()
        {
            return _appConfig.IsFactory;
        }
        /// <summary>
        /// 在程序第一次运行时运行此方法获取配置
        /// </summary>
        /// <returns></returns>
        public InitResult RunInAppStartInit()
        {
            var r = new InitResult() { Result = true };
            if (_seedConfig.IsRemote)//判断是否读取远程配置模式
            {
                ReadRemote(_seedConfig);//读取远程配置
            }
            ReadLocal(_seedConfig, r);//读取本地配置 
            return r;
        }
        public ComponentConfig GetComponentConfig(int componetID)
        {
            //构造函数完成无null初始化设置
            return _appConfig.Components.Find(c => c.ComponentID == componetID);
        }
        public List<ComponentService> GetComponentServices()
        {
            return _appConfig.Workers;
        }
        public SeedConfig GetSeedConfig()
        {
            return _seedConfig;
        }
    }
    internal class InitResult
    {
        public bool Result { set; get; }
        public string Message { set; get; }
    }
}

File: ProgrameSeed/src/main/java/Hayaa/ProgrameSeed/ProgramDistributedConfig.java
package Hayaa.ProgrameSeed;


import java.io.IOException;
import java.net.URLDecoder;
import java.util.HashMap;

import java.util.Map;
import java.util.Properties;
import java.util.UUID;

import org.codehaus.jackson.JsonGenerationException;
import org.codehaus.jackson.map.JsonMappingException;
import org.codehaus.jackson.map.ObjectMapper;

import Hayaa.ProgrameSeed.Config.SeedDefineTable;
import Hayaa.ProgrameSeed.Model.Config.AppConfig;
import Hayaa.ProgrameSeed.Model.Config.ComponentConfig;
import Hayaa.ProgrameSeed.Model.Config.ComponentService;
import Hayaa.ProgrameSeed.Model.Config.SeedConfig;
import Hayaa.ProgrameSeed.Util.FileHelper;
import Hayaa.ProgrameSeed.Util.HttpRequestHelper;
import Hayaa.ProgrameSeed.Util.JsonConvert;
import Hayaa.ProgrameSeed.Util.StringHelper;

public class ProgramDistributedConfig
{
	private AppConfig _appConfig;
	private SeedConfig _seedConfig;	
	private static ProgramDistributedConfig _instance = new ProgramDistributedConfig();

	public static ProgramDistributedConfig Instance()
	{

		return _instance;

	}

	private ProgramDistributedConfig()
	{
		_appConfig = new AppConfig();
		_appConfig.Components = null;
		_appConfig.ConfigContent = "";
		_appConfig.SolutionID = null;
		_appConfig.SolutionName = "解决方案";
		_appConfig.Version = 0;
		try
		{
			_seedConfig = ReadSeedConfig();
		} catch (Exception ex)
		{
			_seedConfig = new SeedConfig();
			_seedConfig.Version = 0;
			_seedConfig.AppConfigID = null;
			_seedConfig.IsRemote = false;
		}
	}

	private void ReadLocal(SeedConfig seedConfig, InitResult result)
	{
		if ((seedConfig.LocalConfigDirectoryPath != null) && (!seedConfig.LocalConfigDirectoryPath.isEmpty()))
		{
			result.Result = false;
			result.Message = "本地配置文件路径为空";
			return;
		}
		try
		{
			AppConfig temp = JsonConvert.DeserializeObject(
					FileHelper.ReadAllText(seedConfig.LocalConfigDirectoryPath + "/" + seedConfig.AppConfigFileName),
					AppConfig.class);
			if (temp == null)
			{
				_appConfig = temp;
			}
		} catch (Exception ex)
		{
			result.Result = false;
			result.Message = ex.getMessage();
			return;
		}
	}

	private AppConfig ReadLocal(SeedConfig seedConfig)
	{
		if (StringHelper.IsNullOrEmpty(seedConfig.LocalConfigDirectoryPath))
		{
			return null;
		}
		try
		{
			AppConfig temp =JsonConvert.DeserializeObject(
					FileHelper.ReadAllText(seedConfig.LocalConfigDirectoryPath + "/" + seedConfig.AppConfigFileName),
					AppConfig.class);
			return temp;
		} catch (Exception ex)
		{

		}
		return null;
	}

	private void ReadRemote(SeedConfig seedConfig) throws JsonGenerationException, JsonMappingException, IOException
	{

		AppConfig localconfig = null;

		if (FileHelper.Exists(seedConfig.LocalConfigDirectoryPath + "/app.Config"))
		{
			localconfig = ReadLocal(seedConfig);
		}
		AppConfig remoteConfig = GetRemote(seedConfig.SeedServerUrl, seedConfig.AppConfigID);

		if (remoteConfig != null)
		{
			if (seedConfig.Version == 0)
			{
				FileHelper.Delete(seedConfig.LocalConfigDirectoryPath + "/" + seedConfig.AppConfigFileName);

				FileHelper.AppendAllText(seedConfig.LocalConfigDirectoryPath + "/" + seedConfig.AppConfigFileName,
						JsonConvert.SerializeObject(remoteConfig));
			}
			if ((seedConfig.Version > 0) && (localconfig == null))
			{
				FileHelper.AppendAllText(seedConfig.LocalConfigDirectoryPath + "/" + seedConfig.AppConfigFileName,
						JsonConvert.SerializeObject(remoteConfig));
			}
		}

	}

	@SuppressWarnings("unchecked")
	private AppConfig GetRemote(String url, UUID solutionID)
	{
		@SuppressWarnings("rawtypes")
		Map<String, String> dic = new HashMap<String, String>();
		dic.put(SeedDefineTable.UrlAction, SeedDefineTable.GetRmoteConfigAction);
		dic.put(SeedDefineTable.SolutionIDParam, solutionID.toString());
		String str = "";
		AppConfig result = null;
		try
		{
			str = HttpRequestHelper.Instance().GetNormalRequestResult(url, dic, "get");
			str = URLDecoder.decode(str, "UTF-8");
			result = JsonConvert.DeserializeObject(str, AppConfig.class);
		} catch (Exception ex)
		{
			result = null;
			return result;
		}
		return result;
	}

	/// <summary>
	///
	/// </summary>
	/// <returns></returns>
	private SeedConfig ReadSeedConfig()
    {
        SeedConfig r = new SeedConfig();
       
             r.IsRemote = true;
             r.IsVirtualPath = true;
            		 r.LocalConfigDirectoryPath = "/Config";
            				 r.Version = 0;
            				 Properties props = System.getProperties();
        try
        {
        	
            String baseDirectory = props.getProperty("java.class.path");
            r=JsonConvert.DeserializeObject(FileHelper.ReadAllText(baseDirectory + "/seed.config"),SeedConfig.class);
        }
        catch (Exception ex)
        {
            r = new SeedConfig(); r.IsVirtualPath = false;
        }
        if (r.IsVirtualPath)
        {
            r.LocalConfigDirectoryPath= props.getProperty("java.class.path")+r.LocalConfigDirectoryPath;
        }
        return r;
    }

	public AppConfig GetLocalConfig()
	{
		return ReadLocal(_seedConfig);
	}

	public Boolean IsEmpty()
	{
		if (_appConfig.SolutionID==null)
		{
			return true;
		}
		return false;
	}

	public Boolean IsFactory()
	{
		return _appConfig.IsFactory();
	}
	
	public InitResult RunInAppStartInit() throws JsonGenerationException, JsonMappingException, IOException
	{
		InitResult r = new InitResult();
		r.Result = true;
		if (_seedConfig.IsRemote)
		{
			ReadRemote(_seedConfig);
		}
		ReadLocal(_seedConfig, r);
		return r;
	}

	public ComponentConfig GetComponentConfig(int componetID)
    {       
		for(int i=0;i<_appConfig.Components.length;i++) {
			if(_appConfig.Components[i].ComponentID == componetID)
			return _appConfig.Components[i];
		}
		return null;
    }

	public ComponentService[] GetComponentServices()
	{
		return _appConfig.CompeontInstances;
	}

	public SeedConfig GetSeedConfig()
	{
		return _seedConfig;
	}
}
